### DISCUSSION TITLE
Understanding `next build` and the deployment process

### DISCUSSION THREAD

**Discussion by BenjaminWFox** [vercel/next.js#14339] [Help] [answered]
Hi all, I'm wondering if someone can help me "think in Next.js" around the process of build & deployment. My question is, I think, somewhat similar to this [unanswered StackOverflow question](https://stackoverflow.com/questions/59782255/how-can-i-make-a-minimal-deployment-of-a-next-js-non-static-app). The whole process feels very different from what I'm used to in the "CSR" world, and [the documentation](https://nextjs.org/docs/deployment) feels like it's telling me to "just use Vercel and trust the magic," which isn't an option in my current case.

To compare the steps I'm thinking about:

CSR:
1. Develop locally
2. `npm run build` when ready to produce, for example, a `dist/`
3. publish `dist/` folder to wherever

Next.js:
1. Develop locally
2. `next build` when ready to produce `.next/`
3. publish _the entire project directory_ to wherever
4. `next start`

My main confusion stems from post-build step 3 (with next.js). If it helps, I am not using a custom server and the deployment, CI/CD, and hosting is all within the Azure ecosystem.

Step 3 (in my case), broken down, looks like:
1. Check code in to main branch (github)
2. Build pipeline kicks off, runs `install` and `build` steps (azure devops)
3. Build pipeline publishes to Release pipeline (azure devops)
4. Release pipeline publishes to App Service (azure devops / portal)

In steps 3 & 4, we end up moving around all of our development code, node_modules, test code, etc. It seems like a lot of extra overhead.

Questions:
- Am I missing something obvious in the way I'm thinking about/approaching any of this?
- Is the best option to just zip the project directory and not think about any non-production files that get pushed around along with it?
  - Can I instead publish only a subset of the project directory that is actually required? (and what subset would that be?)
- Is there a particular reason the `build` step doesn't product a self-contained deployment?

**Comment by jamesmosier** [comment, marked as answer]
It sounds like you are on the right track, but you shouldn't need to include all of your other project files. The bare minimum I usually include are: `next.config.js`, `node_modules`, `package.json`, and `.next` folder. The rest don't need to be copied.

My process running on CI sounds similar to yours:

1. Clone repo in CI
2. Install node_modules
3. `npm run build`
4. Copy `next.config.js`, `node_modules`, `package.json`, and `.next` to the production server
5. Once on the server... `npm run start`


The reason it isn't just a self contained bundle/folder that you just drop on a server is because at the end of the day it is a Node app and not just a client side app. You shouldn't need to bring along all of your development files though, just a select few files required to power the "Node" app itself. 

That's just how I approach it, I'm sure there are many different ways to look at it though.

**Reply by BenjaminWFox** [reply]
Thanks for the reply, that's really helpful! It is definitely harder for me to separate out the responsibilities of Next compared to Node having done relatively little node development. I'll play around with your suggestions!

A couple follow up questions:
- Do you copy node_modules as-is or zip it first (or zip together all the files to transfer)? I had tried copying it w/o zipping but it felt like it took much longer. I'd assumed that was overhead from all the i/o, but that may have been something else in the setup at the time (and I may have been running it on a very small vm 🤔).
- Beyond the bare minimum are the others you tend to copy?
  - If I'm thinking about it right, it seems like at least the `public/` (or `static/` or whatever you might call it) folder would would be copied?

**Reply by jamesmosier** [reply]
* The way we deploy it in particular is zipping up everything we need first. I'd imagine not zipping that big of a folder would be problematic for sure when trying to transfer. 
* And yes, if the project is using `public` (or `static`) folders then those get copied too on my projects. Other than that I don't bring anything else along.

And for what it's worth, I do have other projects where I clone the whole repo and perform the npm install and build on the server itself. But that's just a preference thing (for that project) and it works well for that situation.

**Reply by justincy** [reply]
I can't imagine that picking what to copy is worth the time as long as you're copying the entire `node_modules` folder. In terms of space, I'm sure it's the largest by many orders of magnitude.

**Reply by ThePiyushAggarwal** [reply]
We copy `next.config.mjs`, `public`, `package.json`, `package-lock.json` and `.next` (cache folder inside should be removed).
Then, a `npm install` on the server and `npm start`.
Actually, most of it is done by docker.

**Comment by criess** [comment]
It also seems like a large overhead to me. In my organization the situation is even worse when we have a mono-repo setup with lerna that creates a massive node_modules (for all projects) (721MiB 83K files) that has symlinks inside. 

Maybe there is a tool to create a package.json from an already built nextjs application to only install the minimal server runtime required packages… 

Any advise on such a lerna + nextjs situation would be appreciated.

**Comment by r-s-karpenko** [comment]
Agree with @criess, it's very large overhead to copy entire node_modules, is it possible to copy only runtime required packages?

**Reply by janosgats** [reply]
Running `npm prune` should clear the unnecessary dependencies if you run it before prod deploy: https://docs.npmjs.com/cli/v7/commands/npm-prune

**Comment by soasada** [comment]
I'm following the approach proposed by @jamesmosier but when I execute `npm run start` in the server I'm getting the following error:

```
> frontend@ start /home/frontend
> next start

internal/modules/cjs/loader.js:892
  throw err;
  ^

Error: Cannot find module '../build/output/log'
Require stack:
- /home/frontend/node_modules/.bin/next
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/home/frontend/node_modules/.bin/next:3:35)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/home/frontend/node_modules/.bin/next' ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! frontend@ start: `next start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the frontend@ start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/.npm/_logs/2022-01-15T13_53_05_983Z-debug.log
```
 
Any ideas?

**Reply by s1nuh3** [reply]
This might be too late, but I found that a solution is to use the complete path to the next command, use it like this on your package.json:

```
"scripts": {
  "start": "node_modules/next/dist/bin/next start -p $PORT"
}
```

Or if your'r using Azure Pipelines / Azure App Services:

Option 1: On the pipelines configuration for release deployment on the task for deployment provide the same start command:
`node_modules/next/dist/bin/next start -p $PORT`

Option 2: On the App Service management page go to - > Configuration -> General Settings -> Startup Command and set:
`node_modules/next/dist/bin/next start -p $PORT`

**Comment by manentai** [comment]
Apologies for the probably silly question, but would it be advisable to leave all in the same folder, run `yarn next start` and just have Nginx  reverse proxy on the app port? Maybe using pm2 to keep the app up?

**Comment by MushuPlatypus** [comment]
When running a npm prune --production I cannot start Next
`Error: failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
Error: Cannot find module '@next/bundle-analyzer'`

**Reply by marko-hologram** [reply]
If you installed `@next/bundle-analyzer` as a development dependency, this is something that can happen. It says in the Readme of the project https://github.com/vercel/next.js/tree/canary/packages/next-bundle-analyzer#installation

> Note: if installing as a devDependency make sure to wrap the require in a process.env check as next.config.js is loaded during next start as well.

This can easily be avoided with something like this:

```js
const withBundleAnalyzer = process.env.BUNDLE_ANALYZE
  ? require("@next/bundle-analyzer")({
      enabled: true,
    })
  : (x) => x;
```

If `BUNDLE_ANALYZE` environment variable is not set, this will just be a function that returns whatever was passed to it. Seems to work fine for me 😄 
You can update this check to whatever you want.

**Comment by slhck** [comment]
In the meantime I think there's a _much_ smoother process for deployment that uses [output file tracing](https://nextjs.org/docs/advanced-features/output-file-tracing). It will generate a `.next/standalone` directory that contains all the needed Node packages (but nothing more) and a single startup `server.js` script.

Copy this and the `public` folder to your server, then simply run `node server.js`.

The whole process is shown in this Dockerfile: https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile

**Reply by marko-hologram** [reply]
Yeah "standalone" output is great, but unfortunately it didn't work for me deploying on Windows IIS with IISNode. Using default config that points to `server.js` and tells IISNode to handle that file doesn't work with this configuration because the app isn't started. Or at least it didn't work for me and what I've tested.

We had custom server.js with express because we needed it to deploy to Azure in this configuration and new server.js generated by Next.js just didn't work the same, even if nothing else changed in IIS web.config file.

What we ended up doing was overwriting that "standalone" output server.js with our custom server.js which we used output file tracing on with this package that Next.js uses under the hood https://github.com/vercel/nft.

This is roughly what we have done:
- build the project with regular `next build`
- create our custom server.js file with a node script
- trace dependencies for our newly created custom server.js file with another custom node script we run
- copy those dependencies into standalone output folder created by Next.js build process
- copy all compiled Next.js dependencies because for some reason we couldn't properly trace those (something Webpack related)
- copy that new server.js file to overwrite server.js generated by Next.js
- deploy as before

We do lose some optimization with this process, but we had to use custom server.js anyway before so we were already losing that optimization. What we gained are faster CI deployments because we don't need to zip and deploy tens of thousands of files in node_modules and instead we get a more optimized deploy that is way faster. We could get an ever more slimmed down build if we didn't have to use a custom server, but gains were already huge with this.

It was hard to get to this point because we just ran into issue after issue, but I guess it's just a combination of things and specific deploy requirements we had. Would love to transfer the app to Linux, but that just wasn't possible at this point because of some other complexities with the overall setup on Azure.

**Comment by tusmenko** [comment]
Seems like a good place to rise another question related to building sequence in Next :)
TL;DR: 
- **Is there any way to actually "compile" the TS code with typing checking and linting, without collecting the page's data?**
- **How to set up the build process to be able to deploy the build, to any environment?**  

Long story long:
For now, all works fine for me in the "next" paradigm in general, but one thing bothers me - it's actually that under the "build" stage we are hiding two actions - **compile** + **build**.

Let's say we have a process similar to GitLab flow, where after the BUILD stage of CI we are operating with Artifacts, as with other Node applications. So, in the end, Artifact is a compiled, analyzed, minified, obfuscated (and whatever else), but working and tested piece of code, ready to be deployed to ANY environment, configured by ENV variables. Like, if I want to test my failing PROD build on the PRE-PROD data set - I should be able just to run the same Artifact, but configured to fetch data from pre-production Backend\API.

This implies a major thing - **it should be Environment-agnostic**,  both from content and configuration POV.

"Next flow" brakes this implication in several points:

1. If you are using NEXT_PUBLIC env. variables - they are "baked" into the js bundle during build time. So the Client Side code will stay configured by the ENV variables it was built with, regardless of where it got deployed.
2. If you have any "Static" pages in your project - they will be included in the bundle during the build phase, based on the data from API, defined during the build stage.

Another limitation, with the same roots:
3. To just ensure that the code is valid in terms of Typing\Linting - you need to actually run the BUILD command, and collect all the data for pages, which may take a significant time when you have lots of pages.

For now, I was able to get a step closer to the expected ENV. agnosticism, addressing those points with: 

- Removing all of the NEXT_PUBLIC variables, and propagating those of them I want to be available on the Client Side via `publicRuntimeConfig` + `next/config `+ `getConfig()`. Feels a bit cumbersome but working fine.
- Reading env variables from SSM (or any other config store of choice) and saving them to `env.production` file.

But for the next two - all I came up with is to define the APIs mock, so the `getStaticPaths` method in my SSG pages returns empty arrays (or just any "default" page to try build on). Alternative - tweak the getStaticPaths to return an empty array based on some ENV variable. So during the BUILD stage locally or CI - I won't generate any pages upfront.

It feels "smelly" because 1) It has a performance trade-off for production, even if not too big -> actual pages are built during the initial page visit.
2) I need to tweak the real code to work around conceptual limitations. Meanwhile - The page's code got compiled\linted even though there no page is generated for it during the build stage.

I've also considered running "build" once more, during the deployment stage, right before the `next start` on the actual environment with a real API, and it feels like a compromise, even increasing an instance startup time.

So the open questions are - 
1) Is there any way to actually "compile" the TS code, check typing, and run tests, without collecting the page's data?
2) What are your ways to achieve similar outcomes (i.e. env agnosticism of the build artifact)?

**Reply by olamide226** [reply]
This is a good split of the problem. I think there are really two separate issues here:

1. **Can Next compile/typecheck/lint without collecting static page data?**
2. **Can the final artifact stay environment-agnostic after it has been built?**

REP only tries to solve the second part.

For the `NEXT_PUBLIC_*` problem, the important detail is that those values are not runtime values for the browser. They are copied into the JS bundle during `next build`, so the artifact is already tied to whatever environment existed during CI.

REP takes a different approach:

https://github.com/RuachTech/rep

Instead of this:

```ts
const apiUrl = process.env.NEXT_PUBLIC_API_URL
```

use browser runtime config:

```ts
import { rep } from '@rep-protocol/sdk'

const apiUrl = rep.get('API_URL')
```

Then the same built artifact can be started with different runtime values:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_URL=https://api.preprod.example.com \
  -e REP_PUBLIC_ENV_NAME=preprod \
  my-next-app:latest
```

or:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_URL=https://api.example.com \
  -e REP_PUBLIC_ENV_NAME=production \
  my-next-app:latest
```

The gateway injects those values into the HTML at container startup/request time as inert JSON:

```html
<script id="__rep__" type="application/json">...</script>
```

For a standalone Next.js container, it can sit in front of the Next server:

```sh
NODE_ENV=production node /app/server.js &

exec rep-gateway \
  --mode proxy \
  --port 8080 \
  --upstream localhost:3000
```

So for the environment-agnostic artifact goal, the split becomes:

- Server-only secrets: keep them as normal runtime `process.env` values on the server.
- Browser-visible config: move away from `NEXT_PUBLIC_*` and inject it at runtime with something like REP.
- Static page data: still a separate Next.js build/rendering concern. REP does not remove the need to decide whether a page is static, dynamic, ISR, mocked at build time, etc.

That last point matters: REP is not a replacement for Next’s data fetching model. It just removes one common reason the artifact stops being portable: browser config being baked into the JS bundle during CI.
